### PR TITLE
Resource allocation

### PR DESCRIPTION
**Pull request by jvalderrama**
## DOCKER CPU AND MEMORY ALLOCATION:

https://docs.docker.com/engine/reference/run/#runtime-constraints-on-resources
## APPROACH

The MEMORY will be assigned to control the "max limit of memory in container" in b, k, m or g. Therefore setting the "max memory of limit in the container" means that the processes in the container can use the "max limit memory in the container" assigned in the OpenNebula Templates or CLI and by default the total virtual memory size (–memory-swap) will be set as double of the "max limit of memory in the container" assigned, in this case, MEMORY + SWAP would be 2*max limit of memory in the container, so process can use the same SWAP memory as well.

The CPU will be assigned to attaching containers to cores, it means pin the container's process to a particular processor (core). It can be assigned to pin to a specific core (0), a range of cores (0-3) or some cores (1,2).
## FEATURE DESCRIPTION IN THE FIRST RELEASE

Total swap memory is set as the double of the memory specified for the container
The memory can be assigned in Megabytes by default.
The kernel must have available swap limit capabilities, otherwise the container can take as much memory as it needs (Memory limited without swap)
The CPU can be assigned for cores > 0 and for a specific core.
## HOW TO TEST
1. Install htop http://hisham.hm/htop/ to check memory and CPU workload in the host
   
   sudo apt-get install htop 
2. Create docker container attached to CPU 1 (core1)
   
   onevm create --memory 512 --cpu 1 --disk ubuntu --nic private --net_context
3. Log in into your container and install stress package, stress your container and check with htop. 
   Stress is a simple workload generator for CPU and Memory by Harvard University http://people.seas.harvard.edu/~apw/stress/
   
   apt-get install stress
   
   stress --cpu 6   (Six threads to core 1)
4. Type 'htop' command in the host to see CPU workload in CPU 1 (core1) attached previously to the container
5. Stress memory and check with htop
   
   stress --vm 1 --vm-bytes 500M --vm-hang 0
## KNOW ISSUES FOR THE FIRST RELEASE VERSION AND FUTURE WORK

Can not be assigned for CPU cores:
- The core 0
- A range of cores (0-2) 
- More than one core (0,1)

The above due to https://github.com/OpenNebula/one/blob/master/src/vm/VirtualMachine.cc#L349 , currently it does not support this kind of assignation for CPU values.

Can not be assigned memory in b, k or g due to https://github.com/OpenNebula/one/blob/master/src/vm/VirtualMachine.cc#L341 , currently it does not support "letters" assignation for memory values.

Commonly can appears a warning message as: "WARNING: Your kernel does not support swap limit capabilities, memory limited without swap", it means that limit your memory won't take a fixed limit (memory+swap) because swap is undefined, it will take as much memory as the container needs.

Be sure of the cores available in your host (number of cores) otherwise it can be crash a new container deployment, to check the number of cores in Ubuntu by example type 'grep -c processor /proc/cpuinfo'

### CODE REVIEW

**Comment by dealfonso** [comment]
Please consider revising the documentation that you cite in the text:

https://docs.docker.com/engine/reference/run/#runtime-constraints-on-resources

Check the meaning of the --cpuset-cpus flag as, according to the link that you provide, it does not mean the AMOUNT of cores. It states the identifier of the cores that the container will use. According to the documentation, that means that if you launch 20 containers that use 1 CPU, the result using "--cpuset-cpus 1" for all of them will be that the 20 containers will try to use only the CPU with id "1". If the host had 4 CPUs, the other CPUs (i.e. with identifiers 0, 2 and 3) will not be used by any of the containers.

From the documentation:

```
$ docker run -ti --cpuset-cpus="1,3" ubuntu:14.04 /bin/bash
This means processes in container can be executed on cpu 1 and cpu 3.
```

**Comment by jvalderrama** [comment]
Absolutely agree, it means the identifiers of the core that the container will use...
Therefore the feature will assign to containers the identifier of the core to be used ...
So there will be as many identifiers as cores available in a host ..

**Comment by dealfonso** [comment]
As for my understanding, the code included in this PR is not distributing the cores among the containers. What if I create two consecutive containers with 1 core each? (and I do not create any other container). These two containers will not be limited to use 1 core each. Instead (at most) they will be limited to use 50% of the same core. Is it right?

I think that the objective of this feature is that if I request a VM with 1 CPU, it is restricted to use just 1 core (and not to use core with id 1). And if I request a VM with 4 CPUs, it is restricted to use 4 cores (and not to use just the core with id 4).

**Comment by jvalderrama** [comment]
According with the example, the better explanation is here https://goldmann.pl/blog/2014/09/11/resource-management-in-docker/#_attaching_containers_to_cores 

For the second example there is a restriction, clearly Docker allow to assign more than one core to a container, it means by range (0-2) cores, it means pin the container to cores 0,1,2 or pin a container to specific cores (0,2) it means to cores 0 and 2. But this was specified in the section "KNOW ISSUES FOR THE FIRST RELEASE VERSION AND FUTURE WORK" that will be discuss to allow this, currently it can pin container just to a specific core.

Basically it was the initial approach for the release, otherwise I am open to discuss or change this approach, it means when you mention "I think that the objective of this feature is that if I request a VM with 1 CPU, it is restricted to use just 1 core ", I don't sure if this is possible, the way the I found is to pin the container to a specific core, however I am going to check whether is possible to assign this randomly to use any core but limited to just one but this change the initial approach suggested.

**Comment by dealfonso** [comment]
I think that the proposed approach is not correct as it is implemented in this PR. According to the docker documentation and my experiments, there are two simple ways to divide the CPU between the different containers:

## 1. Using the cpu-shares mechanisim according to the next code:

CPU_SHARE=$[ $CPU \* $[ 1024 / $(cat /proc/cpuinfo | grep processor | wc -l) ] ]
CPU_SHARE_VALUE=$(( $CPU_SHARE > 1024 ? 1024 : $CPU_SHARE ))
docker run ... --cpu-share $CPU_SHARE_VALUE

It is important to take into account the upper bound of the parameter (1024) that will be reached in the case of overcommiting. If we do not take into account this upper limit, it will lead to to avoid errors from docker.

The issue with this method is that when one container is running alone, it will use all the available CPU, but if some other containers are running in the same host, the CPU will be distributed according to the relative values of the cpu-shares.

This mechanism will allow cpu-bursting: i.e. allowing containers to use more cpu than requested, if there are not any other container that needs it.

## 2. Scheduling cores by using the cpuset-cpu mechanism

Other mechanism will consist in assigning the individual cores to the containers and, at the moment of deploying the container, check which of them are used (e.g. using _docker inspect -f '{{.HostConfig.CpusetCpus}}' <container_name>_ command for all the running containers) and assign some of the others to the container that is being deployed.

The issue here is when overcommitting happens: this method will fail or will assign multiple containers to a single core. The problem here is when one of the containers is finished. If it is one of the containers that is sharing a core, there is not any problem, but if it is one of the other, we would have some cores overcommited while there are other that are idle. My advise here is to not to allow overcommiting and to fail in such case.

In this case the cpu will be limited from the moment in which the container is deployed.

**Comment by jvalderrama** [comment]
Ok. I am going to include the first approach "1. Using the cpu-shares mechanisim .." , this can be closer to Amazon EC2 Containers Service http://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#advanced_container_definition_params , therefore I think we need to keep in mind the next:
1. The memory parameter will be mandatory
2. The CPU parameter should be mandatory?
3. Performing tests in my dev environment when the value of CPU is assigned to the VM (Container in OpenNebula), the new VM can be stuck in "PENDING" state due OpenNebula Scheduling process*, because it checks the "host" capacity and if this not fullfill the requirements for the new VM, the new VM will not be created, keeping this state for ever, the message shown for the VM is "No host with enough capacity to deploy the VM", so the point is that it depends of a correct assignation to the user who is going to create the new VM.. 

*https://github.com/OpenNebula/one/blob/master/src/scheduler/src/sched/Scheduler.cc

i.e http://lists.opennebula.org/pipermail/users-opennebula.org/2011-April/032171.html

**Comment by dealfonso** [comment]
Closing because of lack of progress
